manta-crypto/src/arkworks/glv/mod.rs

```diff
@@ -0,0 +1,341 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Arkworks Elliptic Curve Implementation
+
+use crate::arkworks::{
+    ec::{
+        models::{short_weierstrass_jacobian, SWModelParameters},
+        AffineCurve, ProjectiveCurve,
+    },
+    ff::PrimeField,
+};
+use alloc::{vec, vec::Vec};
+use ark_ff::{BigInteger, Field};
+use core::str::FromStr;
+use num_bigint::{BigInt, BigUint, Sign};
+
+#[cfg(feature = "ark-bls12-381")]
+use crate::arkworks::bls12_381;
+
+#[cfg(feature = "ark-bn254")]
+use crate::arkworks::bn254;
+
+/// Affine Curve Extension
+pub trait AffineCurveExt: AffineCurve {
+    /// Returns the `x` coordinate of `self`.
+    fn x(&self) -> &Self::BaseField;
+
+    /// Returns the `y` coordinate of `self`.
+    fn y(&self) -> &Self::BaseField;
+
+    /// Builds [`Self`] from `x` and `y`.
+    fn from_xy_unchecked(x: Self::BaseField, y: Self::BaseField) -> Self;
+
+    /// Applies the GLV endomorphism to `self`
+    #[inline]
+    fn glv_endomorphism(&self, beta: &Self::BaseField) -> Self {
+        Self::from_xy_unchecked(*self.x() * beta, *self.y())
+    }
+}
+
+impl<P> AffineCurveExt for short_weierstrass_jacobian::GroupAffine<P>
+where
+    P: SWModelParameters,
+{
+    #[inline]
+    fn x(&self) -> &Self::BaseField {
+        &self.x
+    }
+
+    #[inline]
+    fn y(&self) -> &Self::BaseField {
+        &self.y
+    }
+
+    #[inline]
+    fn from_xy_unchecked(x: Self::BaseField, y: Self::BaseField) -> Self {
+        Self::new(x, y, false)
+    }
+}
+
+/// Given a scalar `k` and basis vectors `v` and `u` finds integer scalars `k1` and `k2`,
+/// so that `(k, 0)` is close to `k1v + k2u`, meaning the norm of the difference `||(k,0) - (k1v + k2u)||`
+/// is at most `max(||v||, ||u||)`.
+#[inline]
+pub fn decompose_scalar<F>(k: &F, v: (&BigInt, &BigInt), u: (&BigInt, &BigInt)) -> (BigInt, BigInt)
+where
+    F: PrimeField,
+{
+    let k = BigInt::from_bytes_be(Sign::Plus, &k.into_repr().to_bytes_be());
+    let q1 = (u.1 * &k) / ((v.0 * u.1) - (v.1 * u.0));
+    let q2 = (-v.1 * &k) / ((v.0 * u.1) - (v.1 * u.0));
+    let k1 = k - &q1 * v.0 - &q2 * u.0;
+    let k2 = 0 - q1 * v.1 - q2 * u.1;
+    (k1, k2)
+}
+
+#[derive(Clone, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+/// GLV Parameters
```

Comment:
Done, I can see why `Ord` and `PartialOrd` don't make sense. Can you explain what `derivative::Derivative` is for?

Reply:
This is because we are using an associated type inside the structure, and the standard Rust `derive` macro does not have the correct semantics for those. You can always check if the semantics are being done correctly by building the docs and checking the requirements output by the macro and making sure that they are exactly the minimal set, like `C::ScalarField: Clone` vs `C: Clone` which would be wrong.